File: sdk/typescript/rooch-sdk/src/crypto/keypair.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { bech32 } from 'bech32'

import { Signer } from './signer.js'
import type { SignatureScheme } from './signatureScheme.js'
import { SIGNATURE_FLAG_TO_SCHEME, SIGNATURE_SCHEME_TO_FLAG } from './signatureScheme.js'

export const PRIVATE_KEY_SIZE = 32
export const LEGACY_PRIVATE_KEY_SIZE = 64

export const ROOCH_SECRET_KEY_PREFIX = 'roochsecretkey'

export type ParsedKeypair = {
  /**
   * The signature scheme of the private key.
   * @deprecated Use `scheme` instead. This alias is kept for backward-compatibility and will be removed in a future release.
   */
  schema?: SignatureScheme
  /**
   * The signature scheme of the private key.
   */
  scheme: SignatureScheme
  secretKey: Uint8Array
}

export abstract class Keypair extends Signer {
  /**
   * This returns the Bech32 secret key string for this keypair.
   */
  abstract getSecretKey(): string
}

/**
 * Decode a Bech32-encoded Rooch secret key.
 *
 * @param value Bech32 string starting with `roochsecretkey`.
 * @returns ParsedKeypair object containing the secret key bytes and signature scheme.
 */
export function decodeRoochSecretKey(value: string): ParsedKeypair {
  const { prefix, words } = bech32.decode(value)
  if (prefix !== ROOCH_SECRET_KEY_PREFIX) {
    throw new Error('invalid private key prefix')
  }
  const extendedSecretKey = new Uint8Array(bech32.fromWords(words))
  const secretKey = extendedSecretKey.slice(1)
  const signatureScheme =
    SIGNATURE_FLAG_TO_SCHEME[extendedSecretKey[0] as keyof typeof SIGNATURE_FLAG_TO_SCHEME]
  return {
    scheme: signatureScheme,
    schema: signatureScheme, // backward compatibility
    secretKey,
  }
}

/**
 * @deprecated Use {@link decodeRoochSecretKey} instead.
 */
export const decodeRoochSercetKey = decodeRoochSecretKey

/**
 * This returns a Bech32 encoded string starting with `roochsecretkey`,
 * encoding 33-byte `flag || bytes` for the given the 32-byte private
 * key and its signature scheme.
 */
export function encodeRoochSecretKey(bytes: Uint8Array, scheme: SignatureScheme): string {
  if (bytes.length !== PRIVATE_KEY_SIZE) {
    throw new Error('Invalid bytes length')
  }
  const flag = SIGNATURE_SCHEME_TO_FLAG[scheme]
  const privKeyBytes = new Uint8Array(bytes.length + 1)
  privKeyBytes.set([flag])
  privKeyBytes.set(bytes, 1)
  return bech32.encode(ROOCH_SECRET_KEY_PREFIX, bech32.toWords(privKeyBytes))
}

/**
 * @deprecated Use {@link encodeRoochSecretKey} instead.
 */
export const encodeRoochSercetKey = encodeRoochSecretKey

File: sdk/golang/rooch-sdk/crypto/keypair.go
package crypto

import (
	"errors"
	"github.com/btcsuite/btcutil/bech32"
)

const (
	PrivateKeySize       = 32
	LegacyPrivateKeySize = 64
	RoochSecretKeyPrefix = "roochsecretkey"
)

// ParsedKeypair represents a parsed keypair with its signature scheme and secret key
type ParsedKeypair struct {
	Schema    SignatureScheme
	SecretKey []byte
}

// Keypair is an abstract base struct that extends Signer
type Keypair struct {
	Signer
}

// GetSecretKey returns the Bech32 secret key string for this keypair
func (k *Keypair) GetSecretKey() string {
	// This is abstract and should be implemented by concrete types
	panic("GetSecretKey must be implemented by concrete types")
}

// DecodeRoochSecretKey decodes a Bech32 encoded secret key string
func DecodeRoochSecretKey(value string) (*ParsedKeypair, error) {
	// Decode the Bech32 string
	prefix, words, err := bech32.Decode(value)
	if err != nil {
		return nil, err
	}

	if prefix != RoochSecretKeyPrefix {
		return nil, errors.New("invalid private key prefix")
	}

	// Convert words to bytes
	extendedSecretKey, err := bech32.ConvertBits(words, 5, 8, false)
	if err != nil {
		return nil, err
	}

	// Extract flag and secret key
	if len(extendedSecretKey) < 2 {
		return nil, errors.New("invalid secret key length")
	}

	flag := extendedSecretKey[0]
	secretKey := extendedSecretKey[1:]

	// Get signature scheme from flag
	scheme, ok := SignatureFlagToScheme[flag]
	if !ok {
		return nil, errors.New("invalid signature scheme flag")
	}

	return &ParsedKeypair{
		Schema:    scheme,
		SecretKey: secretKey,
	}, nil
}

// EncodeRoochSecretKey encodes a private key and signature scheme to a Bech32 string
func EncodeRoochSecretKey(bytes []byte, scheme SignatureScheme) (string, error) {
	if len(bytes) != PrivateKeySize {
		return "", errors.New("invalid bytes length")
	}

	flag, ok := SignatureSchemeToFlag[scheme]
	if !ok {
		return "", errors.New("invalid signature scheme")
	}

	// Combine flag and private key bytes
	privKeyBytes := make([]byte, len(bytes)+1)
	privKeyBytes[0] = flag
	copy(privKeyBytes[1:], bytes)

	// Encode the combined bytes to Bech32
	encoded, err := bech32.Encode(RoochSecretKeyPrefix, privKeyBytes)
	if err != nil {
		return "", err
	}

	return encoded, nil
} 